mm2src/mm2_main/src/lp_swap.rs

```diff
@@ -682,6 +683,59 @@ pub fn dex_fee_amount_from_taker_coin(taker_coin: &MmCoinEnum, maker_coin: &str,
     dex_fee_amount(taker_coin.ticker(), maker_coin, trade_amount, &dex_fee_threshold)
 }
 
+// This needs discussion. We need a way to determine the watcher reward amount, and a way to validate it at watcher side so
+// that watchers won't accept it if it's less than the expected amount. This has to be done for all coin types, because watcher rewards
+// will be required by both parties even if only one side is ETH coin. Artem's suggestion was first calculating the reward for ETH and
+// converting the value to other coins, which is what I'm planning to do. I based the values to be higher than the gas amounts used
+// when the watcher spends the maker payment or refunds the taker payment. For the validation, I check if the reward is higher
+// than a minimum amount using the min_watcher_reward method. I can't make an exact comparison because the gas price and relative
+// price of the coins will be different when the taker/maker sends their payment and when the watcher receives the message. This should
+// work fine if we pick the WATCHER_REWARD_GAS and MIN_WATCHER_REWARD_GAS constants good. The advantage of this is that the reward will
+// be directly based on the amount of gas burned when the watcher will call the contract functions (Artem's idea was to make it slightly
+// profitable for the watchers). The disadvantage is the comparison during the validations using a separate minimum value. If we want
+// validations with exact values, there are two other ways I could think of:
+// 1.  Precalculating fixed rewards for all coins. The disadvantage is that the gas price and the relative price of coins will change over
+// time and the reward will deviate from the actual gas used by the watchers, and we can't keep updating the values.
+// 2. Picking the reward to be a percentage of the trade amount like the taker fee. The disadvantage is it will be extremely hard to
+// pick the right ratio such that it will be slightly profitable for watchers.
+pub async fn watcher_reward_amount(
+    coin: &MmCoinEnum,
+    other_coin: &MmCoinEnum,
+    watcher_reward: bool,
+) -> Result<Option<u64>, String> {
+    const WATCHER_REWARD_GAS: u64 = 100_000;
+    watcher_reward_from_gas(coin, other_coin, WATCHER_REWARD_GAS, watcher_reward).await
+}
+
+pub async fn min_watcher_reward(
+    coin: &MmCoinEnum,
+    other_coin: &MmCoinEnum,
+    watcher_reward: bool,
+) -> Result<Option<u64>, String> {
+    const MIN_WATCHER_REWARD_GAS: u64 = 70_000;
+    watcher_reward_from_gas(coin, other_coin, MIN_WATCHER_REWARD_GAS, watcher_reward).await
+}
```

Comment:
Wouldn't be better checking `watcher_reward` in `maker/taker_swap.rs` and avoid awaiting these(`watcher_reward_from_gas`, `min_watcher_reward`, `watcher_reward_amount`) `async` functions? This way we can avoid unnecessary context switches, thread yielding, etc. I think it matters for low power devices like mobile phones.

Reply:
Good point. I was trying to avoid repeating this block every time we call these functions:

```rust
let reward = if watcher_reward {
    watcher_reward_amount(taker_coin, maker_coin)
} else {
    None
}
```

Can you think of a better way for this?

Reply:
You can use `then_some` on `watcher_reward`.

```rs
let reward = watcher_reward.then_some(watcher_reward_amount().await?);
```

Reply:
That's what I tried first but `then_some` evaluates the passed function whether the bool is true or false

Reply:
Then your suggestion makes more sense